### DISCUSSION TITLE
Identify a FastDDS example topic on ROS2 [13775]

### DISCUSSION THREAD

**Discussion by Nir-Az** [eProsima/Fast-DDS#2467] [Q&A]
Hi,

If I take a FastDDS example (like HelloWorldSharedMem) and run a publisher,
What need to be changed in order for ROS2 to identify this message? (let say see it in `ros2 topic list` command)
Thanks

**Comment by EduPonz** [comment]
Hi @Nir-Az ,

I think you may take [this](https://gist.github.com/EduPonz/bea0edf3e1ac366560eff62cceb5ddf9) as an starting point to get communication going. The idea is to:
1. Transform the ROS 2 msg file into an IDL
2. Use Fast DDS-Gen with `-typeros2` flag to generate the type support
3. Include the new type support into the example

However, for your publisher to appear in the ROS 2 CLI, you'd have to publish data into the `ros_discovery_info` topic, which ROS 2 uses for distributing discovery information. You can read more about the ROS 2 introspection tools [here](https://docs.ros.org/en/foxy/Concepts/About-Command-Line-Tools.html).

**Reply by Nir-Az** [reply]
Thanks @EduPonz ,

I took the example https://github.com/eProsima/Fast-DDS/tree/master/examples/C%2B%2B/DDS/HelloWorldExample
And rerun `fastddsgen` on the example idl file with `-typeros2` flag,
It didn't change anything except some comment of GUID, so I think it did nothing.

1. Does the step of creating a `.msg` file and converting him is necessary? can't I just run it on an idl file?
How `fastddsgen HelloWorld.idl -typeros2` is different than the msg-> idl process?
2. I will read further on the introspection tools.


Thanks

**Comment by Nir-Az** [comment]
Hey @EduPonz ,

I tried doing to "proper" way and to create a .msg file according to the guide you attached.

I created `devices.msg` containing a single line:
string device_name

Now according to the guide I should use fastddsgen to convert this message into an IDL file?
`Generate a ROS 2 compatible IDL from the ROS 2 .msg using Fast DDS-gen`

How Do I do it?
I don't see this kind of option in the `fastddsgen` tool..
If I try `-typeros2` I get this:
```
devices.msg:1:0: error: mismatched input 'string' expecting {'@', 'typedef', 'custom', 'import', 'struct', 'native', 'eventtype', 'enum', 'home', 'exception', 'const', 'valuetype', 'module', 'union', 'abstract', 'typeprefix', 'typeid', 'local', 'interface', 'component', 'bitset', 'bitmask', '@annotation'}
Exception in thread "main" java.lang.NullPointerException
        at com.eprosima.fastdds.fastddsgen.execute(fastddsgen.java:401)
        at com.eprosima.fastdds.fastddsgen.main(fastddsgen.java:1474)
```

Any help?

I also tried using `fastddsgen` with / without the `-typeros2` flag for an OMG style idl file and the output was identical.
What am I missing?

I created also a ticket on fastddsgen repo (https://github.com/eProsima/Fast-DDS-Gen/issues/92) hoping to get an answer quickly so I can progress with it..

Thanks.

**Reply by EduPonz** [reply]
Hi @Nir-Az ,

It seems there have been changes in ROS 2 since I wrote that small guide that break it. We are writing a more complete guide in Fast DDS documentation. We're still working on the text, but it should be enough for you to complete your task (see [here](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/rosbag_capture/rosbag_capture.html), we'll updated it today or tomorrow). The thing is that sadly, the data type and type name need to meet a specific criteria in order for ROS 2 to detect them.
